#90 - Fix algorithm that finds slots for a particular shadow root

treshugart opened the pull request:
#90

treshugart committed:
fix(slotting): Fix algorithm that finds slots local to a shadow root so that it only includes slots

#90

treshugart commented on src/index.js:
```diff
@@ -4,6 +4,9 @@ import getPropertyDescriptor from './util/get-property-descriptor';
 import debounce from 'debounce';
 import getEscapedTextContent from './util/get-escaped-text-content';
 import getCommentNodeOuterHtml from './util/get-comment-node-outer-html';
+import findSlots from './util/find-slots';
```
I didn't want to add more code to `index.js` so I moved `findSlots()` into a module that requires code shared in `index.js`, so I moved those out, too.

treshugart committed:
docs(slotting): Add comment describing why so many arguments are being passed to createTreeWalker().

adevnadia commented on src/util/is-root-node.js:
```diff
@@ -0,0 +1,3 @@
+export default function(node) {
+  return node.tagName === '_SHADOW_ROOT_';
```
I thought we have a constant for that. `defaultShadowRootTagNameUc`

treshugart replied:
We do but it's configurable via the `attachShadow()` options so we'd have to pass that around. Just created https://github.com/skatejs/named-slots/issues/92. I didn't want to do that here because I wanted to keep the PR's separate.

lukebatchelor commented on test/unit/slot/distribution.js:
```diff
@@ -42,4 +42,59 @@ describe('slot/distribution', function () {
+      const slot2 = document.createElement('slot');
+      const test = document.createTextNode('test');
+
+      // Host2 must be set up and then added to host1's shadow root so that
```
For these more complicated setups, do you think a comment like

```
/*
<host1>
    <test></test>
    <_shadow_root>
        <slot></slot>
        <host2>
            <slot></slot>
            <_shadow_root>
                <slot></slot>
            </_shadow_root>
        </host2>
    </_shadow_root>
</host1>
*/
```

might help?

treshugart committed:
fix(ie): Fix createTreeWalker() quirsk in IE.

soneymathew commented:
lgtm

lukebatchelor commented:
LGTM 👍

treshugart committed:
fix(slotting): Node.SHOW_ELEMENT -> NodeFilter.SHOW_ELEMENT

Commit:
test: added some tests to break find-slots

Commit:
style: linting fix

treshugart committed:
fix(findSlots): Use recursive descent function instead of a tree walker to simplify logic and get te

We can expand on this later to use a tree walker once tests pass and we prove it is a problem.

adevnadia commented:
LGTM
